### ISSUE TITLE
rustc does not detect incorrect (generic) enum structure and overflow its stack instead

### ISSUE THREAD

**Issue by aurantiaco-sucus** [rust-lang/rust#100618]
I tried this code:

```rust
enum Foo<T: 'static> {
    Value(T),
    Recursive(&'static Foo<Option<T>>),
}

fn main() {
    let _x = Foo::Value(());
}
```

I expected to see this happen: *rustc refuses to compile this enum and tell me this (definition) does not work*

Instead, this happened: *rustc overflows its stack and the compile process ends abruptly*

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0-nightly (40336865f 2022-08-15)
binary: rustc
commit-hash: 40336865fe7d4a01139a3336639c6971647e885c
commit-date: 2022-08-15
host: x86_64-pc-windows-msvc
release: 1.65.0-nightly
LLVM version: 15.0.0
```

<details><summary>Backtrace</summary>
<p>

```
thread 'rustc' has overflowed its stack
warning: `whlist` (bin "whlist") generated 9 warnings
error: could not compile `whlist`; 9 warnings emitted

Caused by:
  process didn't exit successfully: `rustc --crate-name whlist --edition=2021 src\main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 -C metadata=0d4447f233920318 --out-dir C:\Users\raida\CLionProjects\whlist\target\debug\deps -C incremental=C:\Users\raida\CLionProjects\whlist\target\debug\incremental -L dependency=C:\Users\raida\CLionProjects\whlist\target\debug\deps` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)
```

</p>
</details>

**Comment by aurantiaco-sucus** [comment]
Formerly discussed [here](https://users.rust-lang.org/t/rustc-overflows-its-stack-with-enums/79948/5).

**Comment by steffahn** [comment]
@rustbot label I-crash
@rustbot label A-codegen (presumably)

Trying to [reproduce on Godbolt’s Compiler Explorer](https://godbolt.org/z/63eKddf1b), I noticed that the crash only happens if `-C debuginfo=2` is provided to `rustc`.

**Comment by aurantiaco-sucus** [comment]
Just tested it again. It still happens with this version of toolchain:
`rustc 1.84.0-nightly (03ee48451 2024-11-18)`
